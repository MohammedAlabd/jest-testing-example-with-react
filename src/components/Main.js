import React, { useState, useEffect } from "react";

export default function Main({ logMe }) {
  const [input, setInput] = useState("");
//   const [pokemon, setPokemon] = useState({});

//   useEffect(() => {
//     const randomID = Math.floor(Math.random() * 100) || 1;
//     fetch(`https://pokeapi.co/api/v2/pokemon/${randomID}`)
//       .then((res) => res.json())
//       .then((pokemonData) => {
//         const {
//           name,
//           sprites: { front_default },
//         } = pokemonData;
//         setPokemon({
//           name,
//           front_default,
//         });
//       });
//   }, []);

  const handleClick = () => {
    if (input){
       logMe(input);

    } 
  };
  return (
    <div>
      <h1>{input || "Hello, World"}</h1>
      <input
        onChange={(e) => setInput(e.target.value)}
        type="text"
        placeholder="input"
      />
      <button data-testid="button" onClick={handleClick}>
        Button
      </button>

      {/* <h1>{pokemon.name}</h1>
      <img src={pokemon.front_default} /> */}
    </div>
  );
}
