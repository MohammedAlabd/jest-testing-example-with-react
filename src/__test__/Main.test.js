import React from "react";
import { render, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import Main from "../components/Main";

describe("the Main.js working correctly", () => {
  // global.fetch = jest.fn(() => Promise.resolve({
  //     json : () => Promise.resolve({
  //         name : "test",
  //         sprites:{front_default: "testImage"}
  //     })
  // }))
  describe("it render the correct items", () => {
    it("It renders The Hello world text", () => {
      const tree = render(<Main />);
      const linkElement = tree.getByText("Hello, World");
      expect(linkElement).toBeInTheDocument();
    });
    it("changes the dom correctly", () => {
      const tree = render(<Main />);
      const input = tree.getByPlaceholderText("input");
      expect(input.value).toBe("");
      fireEvent.change(input, { target: { value: "Re:Coded" } });
      expect(input.value).toBe("Re:Coded");
      const linkElement = tree.getByText("Re:Coded");
      expect(linkElement).toBeInTheDocument();
      // expect(fetch).toHaveBeenCalled()
    });

    it("snapshots", () => {
      const component = renderer.create(<Main />);
      let json = component.toJSON();
      expect(json).toMatchSnapshot();
    });
  });

  describe("it doesn't call the function without input", () => {
    it("it doesn't call the function without input", () => {
      const logMe = jest.fn();
      const tree = render(<Main logMe={logMe} />);
      const button = tree.getByTestId("button");
      fireEvent.click(button);
      expect(logMe).not.toHaveBeenCalled();
    });
  });

  describe("it does call the function with input", () => {
    it("it does call the function with input", () => {
      const logMe = jest.fn();
      const tree = render(<Main logMe={logMe} />);
      const input = tree.getByPlaceholderText("input");
      fireEvent.change(input, { target: { value: "Re:Coded" } });
      const button = tree.getByTestId("button");
      fireEvent.click(button);
      expect(logMe).toHaveBeenCalled();
    });
  });

  describe("it does call the function with input", () => {
    it("it does call the function with input", () => {
      const logMe = jest.fn();
      const tree = render(<Main logMe={logMe} />);
      const input = tree.getByPlaceholderText("input");
      fireEvent.change(input, { target: { value: "Re:Coded" } });
      const button = tree.getByTestId("button");
      fireEvent.click(button);
      expect(logMe).toHaveBeenCalled();
      expect(logMe).toHaveBeenCalledWith("Re:Coded");
    });
  });
});
