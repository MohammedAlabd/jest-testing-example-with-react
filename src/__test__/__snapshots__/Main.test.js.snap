// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Main.js working correctly it render the correct items snapshots 1`] = `
<div>
  <h1>
    Hello, World
  </h1>
  <input
    onChange={[Function]}
    placeholder="input"
    type="text"
  />
  <button
    data-testid="button"
    onClick={[Function]}
  >
    Button
  </button>
</div>
`;
